File: src/modules/role/services/role-update-status.service.ts
import { prisma } from '../../../infra/database/prisma.service';
import { NotFoundException } from '../../../shared/exceptions';
import { UserRequest } from '../../../shared/types/user-request';
import { RoleUpdateStatusDto } from '../domain/dto/role-update-status.dto';
import { RoleEntity } from '../domain/entities/role.entity';
import { roleDefaultValidator } from '../validators/role-default.validator';
import { roleReferenceValidator } from '../validators/role-reference.validator';

class RoleUpdateStatusService {
  async execute(
    id: number,
    dto: RoleUpdateStatusDto,
    userRequest: UserRequest,
  ): Promise<RoleEntity> {
    const role = await prisma.role.findFirst({
      where: {
        id: id,
        organizationId: userRequest?.organizationId ?? undefined,
        companyId: userRequest?.companyId ?? undefined,
        deletedAt: null,
      },
    });

    if (!role) throw new NotFoundException('Perfil não encontrado!');

    await roleDefaultValidator(role);

    if (role.reference)
      await roleReferenceValidator(role.reference, userRequest);

    const updatedRole = await prisma.role.update({
      data: {
        status: dto.status,
      },
      where: {
        id: id,
      },
      include: {
        organization: true,
        company: true,
      },
    });

    return new RoleEntity(updatedRole);
  }
}

export const roleUpdateStatusService = new RoleUpdateStatusService();

File: src/modules/role/services/role_update_status_service.py
from django.db.models import Q
from src.shared.types.user_request import UserRequest
from src.modules.role.models import Role
from src.modules.role.domain.dto.role_update_status_dto import (
    RoleUpdateStatusDTO,
)
from src.modules.role.validators.role_default_validator import (
    role_default_validator,
)
from src.modules.role.validators.role_reference_validator import (
    role_reference_validator,
)
from src.shared.exceptions.not_found_exception import NotFoundException


class RoleUpdateStatusService:
    @classmethod
    def execute(self, id: int, dto: RoleUpdateStatusDTO, user_request: UserRequest):

        role = Role.objects.filter(
            Q(id=id)
            & Q(deleted_at=None)
            & (
                Q(organization__id=user_request.organization.id)
                if user_request.organization
                else Q()
            )
            & (Q(company__id=user_request.company.id) if user_request.company else Q())
        ).first()

        if role is None:
            raise NotFoundException("Perfil não encontrado!")

        role_default_validator(role)

        role_reference_validator(role.reference, user_request)

        role.status = dto.status

        role.save()

        return role
